olive/passes/pytorch/sparsegpt.py

```diff
@@ -0,0 +1,194 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+# Based on the original implementation at
+# https://github.com/IST-DASLab/sparsegpt
+# https://arxiv.org/abs/2301.00774
+# -------------------------------------------------------------------------
+import logging
+from typing import Any, Dict, List, Union
+
+import torch
+
+from olive.hardware.accelerator import AcceleratorSpec
+from olive.model import PyTorchModel
+from olive.passes import Pass
+from olive.passes.olive_pass import PassConfigParam
+from olive.passes.pytorch.sparsegpt_utils import (
+    SparseGPTModule,
+    catch_layer_inputs,
+    get_layer_submodules,
+    get_layers,
+    layers_map,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class SparseGPT(Pass):
+    """Run SparseGPT on PyTorch model."""
+
+    _requires_data_config = True
+
+    @staticmethod
+    def _default_config(accelerator_spec: AcceleratorSpec) -> Dict[str, PassConfigParam]:
+        return {
+            "sparsity": PassConfigParam(
+                type_=Union[float, List[int]],
+                description=(
+                    "Target sparsity. This can be a float or a list of two integers. Float is the target sparsity per"
+                    " layer. List [n,m] applies semi-structured (n:m) sparsity patterns. Refer to"
+                    " https://developer.nvidia.com/blog/accelerating-inference-with-sparsity-using-ampere-and-tensorrt/"
+                    " for more details on 2:4 sparsity pattern."
+                ),
+            ),
+            "blocksize": PassConfigParam(
+                type_=int, default_value=128, description="Blocksize to use for adaptive mask selection."
+            ),
+            "percdamp": PassConfigParam(
+                type_=float,
+                default_value=0.01,
+                description="Percentage of the average Hessian diagonal to use for dampening. Must be in [0,1].",
+            ),
+            "min_layer": PassConfigParam(
+                type_=int, default_value=None, description="Prune all layers with id >= min_layer."
+            ),
+            "max_layer": PassConfigParam(
+                type_=int, default_value=None, description="Prune all layers with id < max_layer."
+            ),
+            "layer_name_filter": PassConfigParam(
+                type_=Union[str, List[str]],
+                default_value=None,
+                description="Only prune layers whose name contains the given string(s).",
+            ),
+            "compute_device": PassConfigParam(
```

Comment:
could we use the accelerator_spec.device?

Reply:
I consider using but decided against it since this is not the same as `accelerator_spec.device` which is the device we are targeting for inference. 
Here, it is the device we want to use to run the algorithm on. for example, our target device might be cpu but we can still use gpu for this algorithm. It doesn't affect the output model. Just runs faster on gpu if available.

Reply:
that make sense. would you please add more comments here?

Reply:
Yes sure! Should I add to the description or just code comment for the dev?

Reply:
we can add comments either place and it is up to you. The purpose of the comments is just to describe why we choose different device.

Reply:
Sounds good. Thanks! adding comment above the parameter.

Reply:
Done!